### DISCUSSION TITLE
Does tensorrt-llm optimize the diffusion model

### DISCUSSION THREAD

**Discussion by songh11** [NVIDIA/TensorRT-LLM#2030] [Q&A]
I would like to know if there is a difference in performance between the diffusion model using tensorrt and tensort-llm optimization. Which is more suitable?

**Comment by Advait251206** [comment]
It depends on the model you're optimizing.

TensorRT-LLM is specifically designed for **Large Language Models (LLMs)** and related transformer-based generative models. Its optimizations focus on components such as:

- Efficient attention kernels (e.g., paged attention, FlashAttention).
- KV cache management.
- Continuous/in-flight batching.
- Speculative decoding.
- Beam search and sampling optimizations.
- Tensor and pipeline parallelism for autoregressive decoding.

These optimizations are tailored to the token-by-token generation pattern of LLMs.

Diffusion models, on the other hand, have a very different inference workflow. They repeatedly execute a denoising network (typically a U-Net or DiT) over many diffusion timesteps rather than autoregressively generating tokens. As a result, many of the optimizations provided by TensorRT-LLM are not applicable.

For diffusion models (e.g., Stable Diffusion, SDXL, Flux, etc.), **standard TensorRT** is generally the more appropriate choice. It focuses on optimizing the neural network execution itself—kernel fusion, precision optimizations (FP16/INT8), layer scheduling, and engine optimization—which directly benefits diffusion inference.

So, in general:

| Model Type | Recommended Runtime |
|------------|---------------------|
| Large Language Models (Llama, Mistral, Qwen, etc.) | TensorRT-LLM |
| Diffusion Models (Stable Diffusion, SDXL, Flux, DiT, etc.) | TensorRT |

If your diffusion model contains transformer-based components (such as DiT-based architectures), TensorRT-LLM still isn't typically the intended runtime, since it doesn't provide the diffusion scheduler or timestep execution logic. Standard TensorRT remains the recommended optimization framework.

One question that would help clarify your use case:

- Which diffusion model are you referring to (e.g., Stable Diffusion XL, Stable Diffusion 3, Flux, DiT, or another model)? The best optimization strategy can vary depending on the architecture.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

**GitHub:** https://github.com/Advait251206
